### PR TITLE
chore: use shallow checkout (expect for windows)

### PR DESCRIPTION
**Pull request by emilyrohrbough**
Borrow logic identified committed @jaffrepaul in the services repo.

The CircleCI checkout command does a full repo checkout with all of the history which we do not need. Updated to use a shallow checkout for non-windows workflows.

The improve checkout speed:

| workflow | [before](https://app.circleci.com/pipelines/github/cypress-io/cypress/57921) | [after](https://app.circleci.com/pipelines/github/cypress-io/cypress/57881) | Improvement |
|--|--|--| --|
| linux-x64 | 39s | 4s | -35s
| linux-arm64 | 36s | 4s | -32s
| darwin-x64 | 21s | 4s | -17s
| darwin-arm64 | 1m 3s | 9s | -54s

Windows --> to maintain the ssh+git checkout behavior, additional work is needed to correctly correlate the ssh keys for windows to successfully authenticate. Take this gain for now.

### CODE REVIEW

**Review by jennifer-shehane — APPROVED** [review]

File: .circleci/workflows.yml [hunk 1]
```diff
@@ -485,11 +481,9 @@ commands:
     description: Install Node version matching .node-version
     steps:
       # installing NVM will use git+ssh, so update known_hosts
```
Comment (astone123) [inline .circleci/workflows.yml]: do we still need this comment if we're not going to update known_hosts?

**Review by astone123 — APPROVED** [review]
nice!

File: .circleci/workflows.yml [hunk 2]
```diff
@@ -486,11 +481,9 @@ commands:
...
```
Comment (emilyrohrbough) [inline .circleci/workflows.yml]:
```suggestion
```

File: .circleci/workflows.yml [hunk 3]
```diff
@@ -157,7 +154,7 @@ commands:
           name: Set environment variable to determine whether or not to persist artifacts
           command: |
             echo "Setting SHOULD_PERSIST_ARTIFACTS variable"
-            echo 'if ! [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "publish-binary" && "$CIRCLE_BRANCH" != "lerna-optimize-tasks" && "$CIRCLE_BRANCH" != "feature/experimental-retries" ]]; then
+            echo 'if ! [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "publish-binary" && "$CIRCLE_BRANCH" != "lerna-optimize-tasks" && "$CIRCLE_BRANCH" != "em/shallow-checkout" ]]; then
```
Comment (emilyrohrbough) [inline .circleci/workflows.yml]:
```suggestion
            echo 'if ! [[ "$CIRCLE_BRANCH" != "develop" && "$CIRCLE_BRANCH" != "release/"* && "$CIRCLE_BRANCH" != "publish-binary" && "$CIRCLE_BRANCH" != "lerna-optimize-tasks" ]]; then
```

**Pull request merged**